src/truncate.rs

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) 2017-present, PingCAP, Inc. Licensed under Apache-2.0.
+
+use crate::errors::Error;
+use crate::errors::Error::InvalidArgument;
+use crate::file_pipe_log::{build_file_reader, build_file_writer, LogFileReader, LogFileWriter};
+use crate::file_pipe_log::{LogFd, ReplayMachine};
+use crate::log_batch::LogItemBatch;
+use crate::log_batch::LogItemContent::{Command, EntryIndexes, Kv};
+use crate::memtable::EntryIndex;
+use crate::truncate::TruncateMode::{All, Back, Front};
+use crate::{log_batch, FileBuilder, FileId, LogBatch, LogQueue};
+use std::collections::HashMap;
+use std::fs;
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+use std::sync::Arc;
+
+pub struct TruncateMachine {
+    items: Vec<LogItemBatch>,
+}
+
+#[derive(Clone, Copy)]
+pub enum TruncateMode {
+    Front = 0,
+    Back = 1,
+    All = 2,
+}
+
+impl FromStr for TruncateMode {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s.to_lowercase().as_str() {
+            "front" => Ok(Front),
+            "back" => Ok(Back),
+            "all" => Ok(All),
+            _ => Err(InvalidArgument(format!(
+                "unsupported truncate mode {}, only support 'front', 'back', 'all'",
+                s
+            ))),
+        }
+    }
+}
+
+#[derive(Clone, Copy)]
+pub struct TruncateQueueParameter<'a> {
+    pub queue: Option<LogQueue>,
+    pub truncate_mode: TruncateMode,
+    pub raft_groups_ids: &'a [u64],
+}
+
+impl Default for TruncateMachine {
```

Comment:
Refactor and export the file lock inside "file_pipe_log" module, then lock the directory when creating new truncate machine.